File: Algorithms/Python/MergeSort.py
def merge_sort(arr):
    if len(arr) > 1:
        mid = len(arr) // 2
        first_h = arr[:mid]
        second_h = arr[mid:]

        merge_sort(first_h)
        merge_sort(second_h)

        i = j = k = 0
        while i < len(first_h) and j < len(second_h):
            if first_h[i] < second_h[j]:
                arr[k] = first_h[i]
                i += 1
            else:
                arr[k] = second_h[j]
                j += 1
            k += 1

        while i < len(first_h):
            arr[k] = first_h[i]
            i += 1
            k += 1
        while j < len(second_h):
            arr[k] = second_h[j]
            j += 1
            k += 1


array = [5, 4, 9, 8, 44, 0, 100]
merge_sort(array)
print(array)

File: Algorithms/JAVA/src/implementing_algorithms/MergeSort.java
package implementing_algorithms;

public class MergeSort {
	public void mergeSort(int arr[], int left, int right) {
		if (left < right) {
			int mid = (left + right) / 2;
			mergeSort(arr, left, mid);
			mergeSort(arr, mid + 1, right);

			merge(arr, left, mid, right);
		}
	}

	public void merge(int arr[], int left, int mid, int right) {
		int firstHalf = mid - left + 1;
		int secondHalf = right - mid;

		int firstH[] = new int[firstHalf];
		int secondH[] = new int[secondHalf];

		for (int i = 0; i < firstHalf; i++)
			firstH[i] = arr[left + i];
		for (int i = 0; i < secondHalf; i++)
			secondH[i] = arr[mid + i + 1];

		int i = 0, j = 0, k = left;
		while (i < firstHalf && j < secondHalf) { // Pick the larger value from both halves and place it in correct
													// position
			if (firstH[i] <= secondH[j]) {
				arr[k] = firstH[i];
				i++;
			} else {
				arr[k] = secondH[j];
				j++;
			}
			k++;
		}

		// Meaning at least one of the halves was fully iterated through, continue with
		// the other
		while (i < firstHalf) {
			arr[k] = firstH[i];
			i++;
			k++;
		}

		while (j < secondHalf) {
			arr[k] = secondH[j];
			j++;
			k++;
		}
	}

	public static void main(String[] args) {
		MergeSort myMS = new MergeSort();
		int arr[] = { 5, 4, 9, 8, 44, 0, 100 };

		myMS.mergeSort(arr, 0, arr.length - 1); // Send by reference

		for (int i = 0; i < arr.length; i++)
			System.out.print(arr[i] + " ");
	}
}